tests/test_virtual_functions.py

```diff
@@ -375,3 +375,22 @@ def test_issue_1454():
     # Fix issue #1454 (crash when acquiring/releasing GIL on another thread in Python 2.7)
     m.test_gil()
     m.test_gil_from_thread()
+
+
+def test_issue_1546():
+    # Fix issue #1546 (Python object not kept alive by shared_ptr)
+    somelist = []
+
+    class Derived(m.SharedPtrBase):
+        def __init__(self):
+            super(Derived, self).__init__()
+
+        def f(self, value):
+            somelist.append(value)
+            print("Right one", 42)
+
+    holder = m.SharedPtrHolder(Derived())
+    # At this point, our 'Derived' instance has gone out of scope in Python but
+    # is being kept alive by a shared_ptr inside 'holder'.
+    holder.run(42)
+    assert somelist == [42]
```

Comment:
Is there a way that you can check and ensure that all objects get destroyed when the go out of scope?

See usages of `ConstructorStats` for how to do so - I'm happy to point out more examples!

Reply:
(The thing I'm still not sure of is if this creates a sort of reference cycle, and effectively cannot be GC'd until the interpreter dies...)